pkg/loadbalancers/l4_test.go

```diff
@@ -3042,3 +3042,140 @@ func createVMInstanceWithTag(t *testing.T, fakeGCE *gce.Cloud, name, tag string)
+			includeLoggingAnnotation: true,
+			enableLogging:            false,
+			logConfigExpected:        false,
+		},
+		{
+			desc:                        "Only sample rate provided, logging not enabled",
+			includeSampleRateAnnotation: true,
+			sampleRate:                  "1",
+			logConfigExpected:           false,
+			errorExpected:               false,
+		},
+		{
+			desc:                        "Invalid sample rate provided, logging not enabled",
+			includeSampleRateAnnotation: true,
+			sampleRate:                  "invalid",
+			errorExpected:               true,
+		},
+		{
+			desc:                     "Logging enabled, sample rate not specified",
+			includeLoggingAnnotation: true,
+			enableLogging:            true,
+			logConfigExpected:        true,
+			expectedSampleRate:       1,
+		},
+		{
+			desc:                        "Logging enabled, sample rate provided",
+			includeLoggingAnnotation:    true,
+			enableLogging:               true,
+			includeSampleRateAnnotation: true,
+			sampleRate:                  "0.5",
+			logConfigExpected:           true,
+			expectedSampleRate:          0.5,
+		},
+		{
+			desc:                        "Logging enabled, invalid sample rate",
+			includeLoggingAnnotation:    true,
+			enableLogging:               true,
+			includeSampleRateAnnotation: true,
+			sampleRate:                  "invalid",
+			errorExpected:               true,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.desc, func(t *testing.T) {
+			t.Parallel()
+
+			svc := test.NewL4ILBService(false, 8080)
+			if tc.includeLoggingAnnotation {
+				if tc.enableLogging {
+					svc.Annotations[annotations.L4LBLoggingAnnotationKey] = annotations.L4LBLoggingEnabled
+				} else {
+					svc.Annotations[annotations.L4LBLoggingAnnotationKey] = ""
+				}
+			}
+			if tc.includeSampleRateAnnotation {
+				svc.Annotations[annotations.L4LBLoggingSampleRateAnnotationKey] = tc.sampleRate
+			}
+
+			nodeNames := []string{"test-node-1"}
+			vals := gce.DefaultTestClusterValues()
+			fakeGCE := getFakeGCECloud(vals)
+
+			namer := namer_util.NewL4Namer(kubeSystemUID, nil)
+
+			networkInfo := network.DefaultNetwork(fakeGCE)
+
+			l4ilbParams := &L4ILBParams{
+				Service:         svc,
+				Cloud:           fakeGCE,
+				Namer:           namer,
+				Recorder:        record.NewFakeRecorder(100),
+				NetworkResolver: network.NewFakeResolver(networkInfo),
+			}
+			l4 := NewL4Handler(l4ilbParams, klog.TODO())
+			l4.healthChecks = healthchecksl4.Fake(fakeGCE, l4ilbParams.Recorder)
+
+			if _, err := test.CreateAndInsertNodes(l4.cloud, nodeNames, vals.ZoneName); err != nil {
+				t.Errorf("Unexpected error when adding nodes %v", err)
+			}
+
+			result := l4.EnsureInternalLoadBalancer(nodeNames, svc)
+			if result.Error != nil {
+				if !tc.errorExpected {
+					t.Errorf("Failed to ensure internal loadBalancer, err %v", result.Error)
+				}
+				return
+			}
+			if result.Error == nil && tc.errorExpected {
```

Comment:
```suggestion
			if tc.errorExpected {
```
we return if the error was not nil above, so it has to be nil here